### ISSUE TITLE
Solution for interpreted language dependencies

### ISSUE THREAD

**Issue by tj** [apex/apex#180]
Currently you pretty much have to have a private registry for node etc, and who knows with python. This might be fine for larger teams or products but for small projects or bootstrapping it's annoying.

Serverless uses "components" here, which solves it nicely, but adds another concept:

```
component1
    |__s-component.json
    |__package.json
    |__lib
        |__index.js
    |__module1
         |__s-module.json
         |__s-templates.json
         |__function1
              |__event.json
              |__handler.js
              |__s-function.json
```

The entire component becomes part of the Lambda function. We could do similar by always including ./functions, or we could have some kind of function.json directive (or .apexinclude? haha) that pulls them in. Or of course we could nest some more as well like Serverless does, but if possible I'd like to avoid that.

**Comment by mthenw** [comment]
Another options:

**Option 1**

Put it in function.json under "dependencies" key and accept relative paths, absolute paths, git repos and github urls. During build all that stuff could lands in some dir ("dependencies" by default) inside zip. 

So in node it would look like this (assuming that node_modules is in project root)

```
{
  ...
  "dependencies": ["../../node_modules"]
  ...
}
```

and then 
```
var foo = require ("./dependencies/node_modules/foo");
```

OR

```
{
  ...
  "dependencies": ["../../node_modules"]
  "dependenciesDir": "."
  ...
}
```
and then 
```
var foo = require ("foo");
```

**Options 2**

Make it magical and always include node_modules in every function zip if there is node_modules inside project root + possibility to turn if off in function.json ("disableNodeModule" yay, super ugly :D)

**Comment by tj** [comment]
Hmm yea I kinda like that. In some ways I think having node_modules and similar in the function dirs themselves makes sense, since each one would have its own package.json, so that might be weird

**Comment by 5loop** [comment]
I'm a fan of the /lib directory + a paradigm of independent functions for large projects. 

Seems like this would be streamlined, scalable, and light on configuration. And it would make migrating to Apex very, very easy... just drop your old back-end stuff in /lib (or /common, or whatever name)

Plus maybe an "excludeLib" parameter in function.json :)

**Comment by tobyjoe** [comment]
FWIW, I'm kind of into the hook idea, after chatting with @tj a little bit on Friday in Slack. It keeps this sort of thing out of Apex - which is nice when you think about one goal of Apex being to support other languages than just JS & Py & Java (per Lambda spec). It could get pretty fat-bottomed and messy pretty quickly. 

But - if we had some sample hooks (or even canonical ones) that you could enable by name, we could enforce not only that hooks let you do a lot of non-Apex stuff through Apex, but we could also nudge people toward some good practices via the default hooks. 

For example, running linters and stuff, or finding dead code (every extra operation Lamdba has to do slows down your response time and the window isn't infinite...)

The built-in hooks (convenience hooks, really) could be included by name, and packaged via go-bindata or similar. Wouldn't need a ton. Just stuff to like, include js in project/lib and functions/{name}/lib with some minification and linting and stuff; running gofmt and the like for golang; i dunno...

Similar to how we all probably use git hooks on our projects...

**Comment by tj** [comment]
I'm pretty happy with the webpack/browserify solutions in _examples now. They solve this problem with a few benefits:

- much smaller file size (~40MB -> ~100kb)
- no need to re-install deps with `npm install --production`

The only downside really is the lack of an "out of the box" feel, but a generator could solve that. At least then the choice of webpack or browserify is just convention, not forced by Apex.

**Comment by 5loop** [comment]
It's weird browserifying server-side stuff, but I'm starting to really like the idea behind your setup in _examples :)

Now I'm imagining a next-gen npm that compiles dependencies haha

**Comment by tj** [comment]
It's definitely weird, but the only reasonably simple solution that doesn't involve them would be having something copy only the production modules into the function.

The other weird thing is that you could have a package.json in each function, or one in the project root. Function level would be more "correct" but so far it seems like people are mostly doing one in root.

Looks like serverless has a root package.json and the component package.json https://github.com/serverless/serverless-graphql-blog/tree/master/blog, not sure how I feel about that, works I suppose but it doesn't strike me as super clean either

**Comment by paultyng** [comment]
Considering lambda file size limitations and the ease with which you can run in to package size limits without following an example and just trying to use apex on your own, I really think a generator or some sort of automatic handling in the nodejs plugin should be included.  

It may be worth it for the nodejs plugin to do a production npm install in a temp directory (could manually assemble a package.json and kill it before the upload to save space) and include that in a zip?  Maybe thats too abnormal of a node process though? I'm not sure.

**Comment by tj** [comment]
I think the normal process is to build in CI with `npm install --production`, but in this case I think that's too slow to be a great default.

If we went the copying route we'd have to assume that everyone uses a root-level package.json, but IMO that isn't great because then you're sharing deps between functions.

**Comment by paultyng** [comment]
Yeah, maybe if there was like a .apex folder in the project with a production build of the packages cached to speed up subsequent deploys?  But doing it each time would be pretty annoying.

I'm coming around to the webpack/browserify option, I think I'm going to switch to that as well, simply to optimize artifact size.

**Comment by 5loop** [comment]
@paultyng It looks like Lambda has a limit of 1024 file descriptors. 

http://docs.aws.amazon.com/lambda/latest/dg/limits.html 

There's a 50 mb limit for deployment size, so nodejs users are going to hit the file descriptor limit way before the size limit. So we're basically going to be forced to bundle with webpack / browserify anyway :)

**Comment by tj** [comment]
@5loop AFAIK node doesn't hold open any of the descriptors for require(), so that should be fine

I think my vote so far would be on the "include" or "dependencies" (would like to reserve this for apex projects later), still feels a little hacky but that might be the only
solution since more nested dirs is meh.

**Comment by cortesi** [comment]
Chipping in here, I'd like something like @mthenw's option 1). An example use case is that I use a shared set of JSON schemas pretty extensively to pull up API Gateway configurations, render docs and do validation inside lambdas. Being able to conveniently reference a JSON schema to include along with the lamba package in the function configuration would be super convenient, and would be nicer than another &&-chained hooks to copy. I can think of a lot of other use cases for this too.

**Comment by victorquinn** [comment]
Adding something else to this discussion, we've run into some issues which have been tricky to solve and I think it has bearing on this conversation.

Basically, we have a node app which depends on a couple modules which use C++ under the hood [nan](https://github.com/nodejs/nan) and [bcrypt](https://www.npmjs.com/package/bcrypt).

Problem being, deployments using Apex by default run the `npm install` on my local machine, then zip up and ship off to lambda. However, due to the C++ dependencies and the fact that my local machine is running OS X, I get `Invalid ELF Header` errors from lambda (because those C++ dependencies built on OS X break when run on the Amazon Linux which lambda is built on).

Likewise, babel and browserify seem to balk at node modules with C++ dependencies so those are a no go.

Anyway, I'm not sure if this needs to be tackled as part of this dependency management issue, but I kind of think it does it's currently a huge pain.

As a short term fix, I'm building out my Makefile to copy my `package.json` to Vagrant or an EC2 server running the same version of linux as Amazon uses for lambda, run npm install (which builds the C++ dependencies on the appropriate architecture) and then copy that back into place on my local machine, all from the Apex build hook. It's messy and it's complicated.

**Comment by mkjois** [comment]
@victorquinn I realize it's been a while, but did you end up with any better solutions to your C++ deps problem? I'm running into the same obstacle and I'm thinking about bringing the [Amazon Linux Docker image](http://docs.aws.amazon.com/AmazonECR/latest/userguide/amazon_linux_container_image.html) into play so I can build anywhere that runs Docker.

**Comment by vespertilian** [comment]
@mkjois It's been a while again, if the binary supports `node-pre-gyp` you can pass arguments to specify a binary that is compatible with lambda. 

Example: 

`npm install grpc --target=6.1.0 --target_arch=x64 --target_platform=linux --target_libc=glibc`

You can see a full example here: 

https://github.com/vespertilian/apex-typescript-binary-deps
